File: src/main/java/edu/metrostate/trackingsystem/domain/repositories/DealerRepository.java
package edu.metrostate.trackingsystem.domain.repositories;

import edu.metrostate.trackingsystem.domain.models.Dealer;
import edu.metrostate.trackingsystem.infrastructure.database.IDatabaseContext;
import edu.metrostate.trackingsystem.infrastructure.database.Result;
import javafx.collections.ObservableList;

import java.util.List;

public class DealerRepository {

    private final IDatabaseContext databaseContext;

    public DealerRepository(IDatabaseContext databaseContext) {
        this.databaseContext = databaseContext;
    }

    public List<String> getDealershipIDs() {
        return databaseContext.getDealershipIDs();
    }

    public void toggleAcquisition(String dealershipId) {
        databaseContext.toggleAcquisition(dealershipId);
    }

    public ObservableList<Dealer> getDealers() {
        return databaseContext.getDealers();
    }

    public Result<Boolean> deleteDealer(String dealershipId) {
        return databaseContext.deleteDealer(dealershipId);
    }

    public boolean updateDealer(String dealershipId, String name) { return databaseContext.updateDealer(dealershipId, name); }

    public boolean updateDealer(Dealer dealer, String name) { return databaseContext.updateDealer(dealer, name); }

    public boolean updateDealer(Dealer dealer) { return databaseContext.updateDealer(dealer); }

}

File: src/main/kotlin/edu/metrostate/dealership/domain/repositories/DealerRepository.kt
//Used for transferring data between application layers from the database (domain layer) to the services layer
package edu.metrostate.dealership.domain.repositories

import edu.metrostate.dealership.domain.models.Dealer
import edu.metrostate.dealership.infrastructure.database.DatabaseContext
import edu.metrostate.dealership.infrastructure.database.Result
import javafx.collections.ObservableList

class DealerRepository(private val databaseContext: DatabaseContext) {
    val dealershipIDs: List<String?>?
        get() = databaseContext.getDealershipIDs()

    fun toggleAcquisition(dealershipId: String?) {
        if (dealershipId != null) {
            databaseContext.toggleAcquisition(dealershipId)
        }
    }

    val dealers: ObservableList<Dealer>
        get() = databaseContext.dealers

    fun deleteDealer(dealershipId: String?): Result<Boolean>? {
        return dealershipId?.let { databaseContext.deleteDealer(it) }
    }

    fun updateDealer(dealershipId: String?, name: String?): Unit? {
        return dealershipId?.let {
            if (name != null) {
                databaseContext.updateDealer(it, name)
            }
        }
    }
}